File: PAT_Basic_Level_by_java/1105 链表合并/1105 链表合并.java
/*
 * 1105 链表合并
 * 
 * 题目描述：
 *     给定两个单链表L1和L2，将较短的链表逆序后，按每2个长链表节点
 *     插入1个短链表节点的规则合并到长链表中。
 *     输出合并后链表的每个节点的地址、数据和下一个节点地址。
 *     题目保证长链表长度至少是短链表长度的2倍。
 * 
 * 实现原理：
 *     解题思路：
 *         1. 使用HashMap存储所有节点，以节点地址为键，方便O(1)查找。
 *         2. 从两个链表的头节点开始，分别遍历构建两个有序的Node列表。
 *         3. 比较两个链表长度，确定长链表和短链表。
 *         4. 将短链表逆序。
 *         5. 遍历长链表，每放入2个长链表节点，就插入1个短链表节点。
 *         6. 最后遍历结果列表，更新每个节点的next指针并输出。
 *     时间复杂度：O(N)，空间复杂度：O(N)。
 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Main {
    // 链表节点类，包含地址、数据和下一节点地址
    static class Node {
        String address;
        int data;
        String next;
        Node(String addr, int d, String n) {
            address = addr;
            data = d;
            next = n;
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] firstLine = br.readLine().split(" ");
        String head1 = firstLine[0];  // 第一条链表的头节点地址
        String head2 = firstLine[1];  // 第二条链表的头节点地址
        int n = Integer.parseInt(firstLine[2]);  // 节点总数
        
        // 用HashMap存储所有节点，地址作为键
        Map<String, Node> nodes = new HashMap<>();
        for (int i = 0; i < n; i++) {
            String[] parts = br.readLine().split(" ");
            nodes.put(parts[0], new Node(parts[0], Integer.parseInt(parts[1]), parts[2]));
        }
        
        // 根据头地址遍历构建两个链表
        List<Node> list1 = buildList(head1, nodes);
        List<Node> list2 = buildList(head2, nodes);
        
        // 确定长链表和短链表
        List<Node> longList, shortList;
        if (list1.size() >= list2.size()) {
            longList = list1;
            shortList = list2;
        } else {
            longList = list2;
            shortList = list1;
        }
        
        // 将短链表逆序
        List<Node> reversedShort = new ArrayList<>();
        for (int i = shortList.size() - 1; i >= 0; i--) {
            reversedShort.add(shortList.get(i));
        }
        
        // 合并链表：长链表每2个节点插入一个短链表节点
        List<Node> result = new ArrayList<>();
        int shortIdx = 0;  // 短链表当前索引
        for (int i = 0; i < longList.size(); i++) {
            result.add(longList.get(i));
            // 每放入2个长链表节点后，插入一个短链表节点
            if ((i + 1) % 2 == 0 && shortIdx < reversedShort.size()) {
                result.add(reversedShort.get(shortIdx));
                shortIdx++;
            }
        }
        
        // 如果短链表还有剩余（题目保证长链表至少是短链表的2倍，所以一般不会有剩余）
        while (shortIdx < reversedShort.size()) {
            result.add(reversedShort.get(shortIdx));
            shortIdx++;
        }
        
        // 输出合并后的链表，更新每个节点的next指针指向下一个节点的地址
        for (int i = 0; i < result.size(); i++) {
            Node node = result.get(i);
            String nextAddr = (i == result.size() - 1) ? "-1" : result.get(i + 1).address;
            System.out.println(node.address + " " + node.data + " " + nextAddr);
        }
    }
    
    // 根据头地址遍历链表，将节点按顺序存入List返回
    static List<Node> buildList(String head, Map<String, Node> nodes) {
        List<Node> list = new ArrayList<>();
        String curr = head;
        while (!curr.equals("-1")) {
            Node node = nodes.get(curr);
            list.add(node);
            curr = node.next;  // 移动到下一个节点
        }
        return list;
    }
}

File: PAT_Basic_Level_by_py3/1105 链表合并/1105 链表合并.py
# 题目：1105 链表合并
#
# 题目描述：
#   给定两个单链表 L_1 = a_1 \to a_2\to \cdots \to a_{n-1}\to a_n 和 L_2 = b_1 \to b_2\to \cdots \to b_{m-1}\to b_m。如果 n\ge 2m，你的任务是将比较短的那个链表逆序，然后将之并入比较长的那个链表，得到一个形如 a_1 \to a_2 \to b_{m} \to a_3 \to a_4 \to b_{m-1}\cdots  的结果。例如给定两个链表分别为 6→7 和 1→2→3→4→5，你应该输出 1→2→7→3→4→6→5。
#
# 输入格式：
#   输入首先在第一行中给出两个链表 L_1 和 L_2 的头结点的地址，以及正整数
#   N (\le 10^5)，即给定的结点总数。一个结点的地址是一个 5 位数的非负整数，空地址 NULL 用 `-1` 表示。
#   随后 N 行，每行按以下格式给出一个结点的信息：
#   ```
#   Address Data Next
#   ```
#   其中 `Address` 是结点的地址，`Data` 是不超过 10^5 的正整数，`Next` 是下一个结点的地址。题目保证没有空链表，并且较长的链表至少是较短链表的两倍长。
#
# 输出格式：
#   按顺序输出结果链表，每个结点占一行，格式与输入相同。
#
# 实现原理：
#   - 使用字典(Dict)存储键值对映射
#   - 循环迭代处理
#   - 列表操作
#
# 算法思路：
#   链表处理：使用字典存储节点地址映射，
#   先遍历链表得到节点顺序，
#   再按K个一组进行反转，最后重新连接输出。
#
import sys

h1,h2,n=sys.stdin.readline().split();d={}
for _ in range(int(n)):
 a,x,b=sys.stdin.readline().split();d[a]=(x,b)
def walk(h):
 r=[]
 while h!='-1':r.append(h);h=d[h][1]
 return r
a,b=walk(h1),walk(h2)
if len(a)<len(b):a,b=b,a
out=[]
for i,x in enumerate(a):
 out.append(x)
 if i%2==1 and b:out.append(b.pop())
out+=b[::-1]
for i,x in enumerate(out):print(x,d[x][0],out[i+1] if i+1<len(out) else '-1')
